Add the one-block multi-thread global reduction support.

zheng-xq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36563 Add sync-point insertions and block/thread local memory allocations
* #36480 Make trivial thread-idx for degenerate statements without thread-idx.
* **#36306 Add the one-block multi-thread global reduction support.**
* #36293 Add trivial reduce for Cuda

Missing __syncthreads between sections.

Differential Revision: [D20957254](https://our.internmc.facebook.com/intern/diff/D20957254)

zheng-xq committed:
Add the one-block multi-thread global reduction support.
Missing __syncthreads between sections.

[ghstack-poisoned]

bwasti commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -516,7 +569,14 @@ void CudaCodeGen::Initialize() {
 
+#if 1
   CompileToNVRTC(oss_.str(), func_name);
+#else
```
rm?

bwasti commented on torch/csrc/jit/tensorexpr/stmt.h:
```diff
@@ -515,6 +515,46 @@ class For : public StmtNode<For> {
   LoopOptions loop_options_;
 };
+
+// A Cuda specific IR Node that implements atomic-add.
```
probably useful in a CPU context as well (`lock xadd` in x86)

bwasti reviewed (approved):
Definitely makes sense to have an AtomicAdd primitive.

As a future iteration we can do something like match on `Reduce(+)` and map that to `Pointwise(atomicAdd)` that stores to an output with stride 0.

zheng-xq committed:
Update on "Add the one-block multi-thread global reduction support."


Missing __syncthreads between sections.

[ghstack-poisoned]

zheng-xq committed:
Update on "Add the one-block multi-thread global reduction support."


Missing __syncthreads between sections.

Differential Revision: [D20957254](https://our.internmc.facebook.com/intern/diff/D20957254)

[ghstack-poisoned]

zheng-xq committed:
Update on "Add the one-block multi-thread global reduction support."


Missing __syncthreads between sections.

Differential Revision: [D20957254](https://our.internmc.facebook.com/intern/diff/D20957254)

[ghstack-poisoned]

zheng-xq committed:
Update on "Add the one-block multi-thread global reduction support."


Missing __syncthreads between sections.

Differential Revision: [D20957254](https://our.internmc.facebook.com/intern/diff/D20957254)

[ghstack-poisoned]

zheng-xq committed:
Update on "Add the one-block multi-thread global reduction support."


Missing __syncthreads between sections.

Differential Revision: [D20957254](https://our.internmc.facebook.com/intern/diff/D20957254)

[ghstack-poisoned]

zheng-xq committed:
Update on "Add the one-block multi-thread global reduction support."


Missing __syncthreads between sections.

Differential Revision: [D20957254](https://our.internmc.facebook.com/intern/diff/D20957254)

[ghstack-poisoned]

SplitInfinity commented:
@ZolotukhinM 

I think this stack broke the `clang-format` job. It's not the end of the world, but it's probably a good idea to let everyone else working on this part of the codebase know that you added the `tensorexpr` directory to `clang-format` so they should format their code and pay attention to the job.

ZolotukhinM commented:
> @ZolotukhinM
> 
> I think this stack broke the `clang-format` job. It's not the end of the world, but it's probably a good idea to let everyone else working on this part of the codebase know that you added the `tensorexpr` directory to `clang-format` so they should format their code and pay attention to the job.

I did let people know, but yeah, I forgot to ask everyone to rebase and rerun the CI, sorry.
